### PR TITLE
made a concurrent version that excels  particularly well on big numbers

### PR DESCRIPTION
**Pull request by alexewerlof**

### CODE REVIEW

File: main.go [hunk 1]
```diff
@@ -4,28 +4,47 @@ import (
...
 	nStr := os.Args[1]
-	nParsed := new(big.Int)
-	_, parsingSuccess := nParsed.SetString(nStr, 10)
-	if !parsingSuccess {
-		panic(fmt.Sprintf("Failed to parse %s %T as an integer number", nStr, nStr))
+	nParsed, parsingError := strconv.Atoi(nStr)
+	if parsingError != nil {
+		panic(fmt.Sprintf("Failed to parse %s %T as an integer number because %s", nStr, nStr, parsingError))
 	}
 	fmt.Printf("%d!\n", nParsed)
```
Comment (dcaba) [inline main.go]: this and next line look a bit inconsistent ;)

File: main.go [hunk 2]
```diff
@@ -4,28 +4,47 @@ import (
 	"fmt"
 	"math/big"
 	"os"
+	"strconv"
 )
 
 var big0 = big.NewInt(0)
```
Comment (dcaba) [inline main.go]: are you still using them?

File: main.go [hunk 3]
```diff
@@ -4,28 +4,47 @@ import (
...
+	for i := 0; i < x; i++ {
+		go func() {
+			c := <-count
+			if c == 0 {
+				res <- <-vals
+			} else {
+				a := <-vals
+				b := <-vals
+				count <- c - 1
+				c := new(big.Int)
```
Comment (dcaba) [inline main.go]: are you reusing a variable here? For clarity, I wouldnt do that

File: main.go [hunk 4]
```diff
@@ -4,28 +4,47 @@ import (
...
+	res := make(chan *big.Int)
+	count := make(chan int)
+	for i := 1; i <= x; i++ {
+		j := big.NewInt(int64(i))
+		vals <- j
 	}
-	var res, xMinOne big.Int
-	return res.Mul(x, factorial(xMinOne.Sub(x, big1)))
+	for i := 0; i < x; i++ {
+		go func() {
```
Comment (dcaba) [inline main.go]: i'd consider changing a little bit the approach. In this case, you are starting x goroutines (one per process?). I'd consider just having a few that, in an "infinite" loop,  processes data as soon as they are in the channel (and waits until more data arrives)

File: main.go [hunk 5]
```diff
@@ -4,28 +4,47 @@ import (
...
 
 var big0 = big.NewInt(0)
 var big1 = big.NewInt(1)
 
-func factorial(x *big.Int) *big.Int {
-	if x.Cmp(big0) == 0 {
-		return big1
+func factorial(x int) *big.Int {
+	vals := make(chan *big.Int, x+1)
+	res := make(chan *big.Int)
```
Comment (dcaba) [inline main.go]: if you don't specify the size of the channel, probably some goroutines will wait (maybe that's the desired behavior)

**Comment by dcaba** [comment]
Hi. I didn't try to understand the logic (to see if its actually valid), but added some considerations maybe you want to revisit (there's a chance the code is right and its already a result of making conscious decisions ;) ). In any case, very good job considering this is your first golang code!!!! Enjoy Christmas in the north ;) , and best regards!

File: main.go (hunk 2 above)
Comment (alexewerlof) [reply main.go]: yes, in vscode go fmt runs on every save so it's almost impossible to be sloppy. the default settings have a habit of removing unused imports automatically which is frustrating at times! :smile:

File: main.go (hunk 3 above)
Comment (alexewerlof) [reply main.go]: nice catch! Will fix. :+1:

File: main.go (hunk 4 above)
Comment (alexewerlof) [reply main.go]: good idea :+1:

File: main.go (hunk 5 above)
Comment (alexewerlof) [reply main.go]: yes that one is intentional. `res` (standing for *result*) makes the final `return <- res` wait till the result is ready.

**Pull request merged**